### PR TITLE
Move content build kickoff to CMS

### PR DESCRIPTION
**Pull request by cweagans**
## Description

Ref #8108

This is very much a WIP. Search for TODO to see some of the stuff left on my radar. Additionally, documentation will be needed.

This is massive and I'm sorry about it.

Things that would help me review-wise right now:
* Do you understand how the process of requesting a frontend build works? Is it clear?
* Are there other things that would increase the reliability of dispatching a build?
* Are there other things that could be moved into/out of the build trigger module? I'm trying to keep that module hyper focused on things related to dispatching and reporting on content releases. One thing that's already on my radar (for a different ticket) is to move the environment discovery stuff out of the build trigger module and generalize it a bit (ticket is https://github.com/department-of-veterans-affairs/va.gov-cms/issues/7727)

## QA steps

What needs to be checked to prove this works?  
What needs to be checked to prove it didn't break any related things?  
What variations of circumstances (users, actions, values) need to be checked?  

As user _uid_ with _user_role_
1. Do this
   - [ ] Validate that
2. Then
   - [ ] Validate that
3. Then validate Acceptance Criteria from issue
   - [ ] a
   - [ ] b
   - [ ] c

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [x] `Platform CMS Team`
- [ ] `Sitewide program`
- [ ] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public websites`
- [ ] `⭐️ Facilities`
- [ ] `⭐️ User support`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

File: .ddev/mutagen/mutagen.yml [hunk 1]
```diff
@@ -11,17 +12,26 @@ sync:
...
+      # into the container with a traditional docker bind-mount
+      - "/.git"
+
       - "/.tarballs"
       - "/.ddev/db_snapshots"
       - "/.ddev/.importdb*"
       - ".DS_Store"
       - ".idea"
+      
+      - "/docroot/sites/default/files"
```
Comment (ndouglas) [inline .ddev/mutagen/mutagen.yml:25]: Is this leading slash correct?

File: .ddev/commands/web/phpcbf [hunk 2]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
```
Comment (timcosgrove) [inline .ddev/commands/web/phpcbf:1]: Is this or any of the `.ddev/` work specific to the work being done for #8108?

File: docroot/modules/custom/va_gov_build_trigger/src/Environment/EnvironmentPluginBase.php [hunk 3]
```diff
@@ -61,23 +51,30 @@ public static function create(ContainerInterface $container, array $configuratio
...
     );
   }
 
   /**
    * {@inheritDoc}
    */
-  public function getWebUrl(): string {
-    return Settings::get('va_gov_frontend_url') ?? 'https://www.va.gov';
+  public function triggerFrontendBuild() : void {
+    $this->filesystem->saveData('build plz', 'public://.buildrequest', FileSystemInterface::EXISTS_REPLACE);
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_build_trigger/src/Environment/EnvironmentPluginBase.php:62]: Assume placeholder?

File: docroot/modules/custom/va_gov_build_trigger/src/Plugin/AdvancedQueue/JobType/ReleaseDispatch.php [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    switch ($this->releaseStateManager->canAdvanceStateTo(ReleaseStateManager::STATE_DISPATCHED)) {
+      case ReleaseStateManager::STATE_TRANSITION_OK:
+        try {
+          $this->environmentDiscovery->triggerFrontendBuild();
+        }
+        catch (\Exception $e) {
+          return JobResult::failure('Release dispatch failed with error: ' . $e->getMessage());
+        }
+        $this->releaseStateManager->advanceStateTo(ReleaseStateManager::STATE_DISPATCHED);
+        $message = 'Content release has been dispatched.';
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_build_trigger/src/Plugin/AdvancedQueue/JobType/ReleaseDispatch.php:109]: Are these only logged or are they displayed to editors? If they are displayed to editors, has there been any discussion w/ UX about the language?

File: docroot/modules/custom/va_gov_build_trigger/src/Plugin/Block/ContentReleaseStatusBlock.php [hunk 5]
```diff
@@ -8,6 +8,16 @@
 use Drupal\Core\Url;
 use Symfony\Component\DependencyInjection\ContainerInterface;
 
+// @todo redo this to use the release state stuff
```
Comment (timcosgrove) [inline docroot/modules/custom/va_gov_build_trigger/src/Plugin/Block/ContentReleaseStatusBlock.php:11]: Are these todo's for this PR or other tickets?

**Review by timcosgrove — COMMENTED** [review]
More of a surface review, basic questions

File: .ddev/commands/web/phpcbf (hunk 2 above)
Comment (cweagans) [reply .ddev/commands/web/phpcbf:1]: No, I can split that off. I just needed it to maintain my sanity. :)

File: docroot/modules/custom/va_gov_build_trigger/src/Environment/EnvironmentPluginBase.php (hunk 3 above)
Comment (cweagans) [reply docroot/modules/custom/va_gov_build_trigger/src/Environment/EnvironmentPluginBase.php:62]: Not a placeholder. The contents of the file don't matter. `build-frontend.sh` just checks to see if that file exists.

File: docroot/modules/custom/va_gov_build_trigger/src/Plugin/AdvancedQueue/JobType/ReleaseDispatch.php (hunk 4 above)
Comment (cweagans) [reply docroot/modules/custom/va_gov_build_trigger/src/Plugin/AdvancedQueue/JobType/ReleaseDispatch.php:109]: Only logged. I'd like to talk about how to present any of this information to end users.

File: docroot/modules/custom/va_gov_build_trigger/src/Plugin/Block/ContentReleaseStatusBlock.php (hunk 5 above)
Comment (cweagans) [reply docroot/modules/custom/va_gov_build_trigger/src/Plugin/Block/ContentReleaseStatusBlock.php]: There are a couple for this one, but I'll create separate tickets for non critical ones.

**Comment by ndouglas** [comment]
tests

**Comment by cweagans** [comment]
Tests are coming today.😅

File: .ddev/commands/web/phpcbf (hunk 2 above)
Comment (cweagans) [reply .ddev/commands/web/phpcbf:1]: This and related changes have been submitted in a separate PR: https://github.com/department-of-veterans-affairs/va.gov-cms/pull/9036

**Comment by ndouglas** [comment]
@cweagans I made some changes in [my PR ](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/9101) that might be worth bringing in before it's actually accomplished its mission.  I encountered errors because of GitFactory's requests for some services in `va_gov_build_trigger`, so I refactored to replace them.  You might've already dealt with it in a different way, or might not have encountered that problem at all, IDK.

**Comment by cweagans** [comment]
That service may end up being unused after this. I haven't ventured too far outside of the `va_gov_build_trigger` module here. The web build stuff was removed in this PR, so it's possible that any errors you were seeing in `GitFactory` were introduced in this PR.

**Comment by cweagans** [comment]
(Also, just as a heads up - I have held off on rebasing or updating this PR so as not to throw off your PR. After your PR is merged, I can rebase this to account for some of the standalone things I pulled out of this and merged separately)

**Comment by ndouglas** [comment]
Don't hold off on my account 🙂

**Comment by ndouglas** [comment]
`settings.ddev.php` relies on the `DDEV_APPROOT` env var to set `$settings['va_gov_app_root']` and `$settings['va_gov_web_root']`, but that variable's not set inside containers.  I added a [commit](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/9101/commits/0bdf9166045e6c0bb6295f75ebdf5e73aecc33f4) to set it, but you might have a better approach in mind.

**Comment by ndouglas** [comment]
I logged in as a `content_admin` user in DDEV and clicked the "Release content" button on the page.  The page transparently refreshed and no other information was provided.  Checking the log messages, I saw this:

Screen Shot 2022-05-24 at 10 06 56 AM

**Comment by cweagans** [comment]
That rebase was _awful_. 😂 😭 

> settings.ddev.php relies on the DDEV_APPROOT env var to set $settings['va_gov_app_root'] and $settings['va_gov_web_root'], but that variable's not set inside containers. I added a [commit](https://github.com/department-of-veterans-affairs/va.gov-cms/pull/9101/commits/0bdf9166045e6c0bb6295f75ebdf5e73aecc33f4) to set it, but you might have a better approach in mind.

That's a bummer. I thought there was some existing env var that had this, but it's probably also fine to manually set it like that.

> I logged in as a content_admin user in DDEV and clicked the "Release content" button on the page. The page transparently refreshed and no other information was provided. Checking the log messages, I saw this:

The permission probably needs assigned to the role. I'll poke at it tomorrow.

File: docroot/modules/custom/va_gov_build_trigger/src/RouteProvider.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      $route = new Route(
+        '/api/content_release/' . $state,
+        // Defaults.
+        [
+          '_controller' => ContentReleaseNotificationController::class . '::' . $state,
+        ],
+        // Requirements.
+        [
+          '_permission' => 'access bulletin queue trigger api',
+          '_user_is_logged_in' => TRUE,
```
Comment (ndouglas) [inline docroot/modules/custom/va_gov_build_trigger/src/RouteProvider.php:39]:
```suggestion
          '_user_is_logged_in' => 'TRUE',
```

**Review by ndouglas — DISMISSED** [review]
I went through this pretty much line-by-line yesterday (Tuesday, June 7) and this morning (Wednesday, June 8).  I think the important changes are already adequately covered by tests.  I haven't discovered any issues other than those I've already pointed out in comments or my PRs for #9015 and #9016.  IMHO we should shift to the manual testing and move to merge this sooner, rather than later, to avoid further painful rebases.  Great work!

**Comment by ndouglas** [comment]
I mean, don't get me wrong, we could always use more tests, but I was reading through and thinking, "oh, yeah, that's a lot, we should have tests for that" and then I'd flip over and see that you already wrote them.

File: tests/phpunit/FrontendBuild/ReleaseStateManagerTest.php [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+
+  /**
+   * The event dispatcher service.
+   *
+   * @var \Symfony\Component\EventDispatcher\EventDispatcherInterface
+   */
+  protected $eventDispatcher;
+
+  /**
+   * The state service.
```
Comment (ndouglas) [inline tests/phpunit/FrontendBuild/ReleaseStateManagerTest.php:27]: This comment feels so angry, lol.

Comment (cweagans) [reply tests/phpunit/FrontendBuild/ReleaseStateManagerTest.php:27]: i was irritated by our phpcs/phpstan stuff 😂

**Comment by ElijahLynn** [comment]
cypress failure is the content_release.feature:

```
Stderr:                                                     
Command "va-gov-build-frontend-empty-queue" is not defined. 
```

![image](https://user-images.githubusercontent.com/1504756/174355328-0c18d31f-6b13-42bb-a8ee-7a7db7b67e22.png)


```
❯ tugboat log 626c80d40356a744ff1c4c58 | grep -A 100 "Running:  behavioral/content_release.feature"                                                                                        
  Running:  behavioral/content_release.feature                                             (2 of 17)                                                                                       
                                                                                                                                                                                           
                                                                                                                                                                                           
  Content Release                                                                                                                                                                          
    1) The content release page should normally display no in-process releases                                                                                                             
    2) The content release page should show a pending default release initiated within the browser                                                                                         
    3) The content release page should show a pending chosen release initiated within the browser                                                                                          
    4) The content release page should show a pending default release initiated from the command line                                                                                      
    5) The content release page should show a pending chosen release initiated from the command line                                                                                       
                                                                                                                                                                                           
                                                                                                                                                                                           
  0 passing (28s)                                                                                                                                                                          
  5 failing                                                                                                                                                                                
                                                                                                                                                                                           
  1) Content Release
       The content release page should normally display no in-process releases:
     CypressError: `cy.exec('drush va-gov-build-frontend-empty-queue')` failed because the command exited with a non-zero code.

Pass `{failOnNonZeroExit: false}` to ignore exit code failures.

Information about the failure:
Code: 1

Stderr:
Command "va-gov-build-frontend-empty-queue" is not defined.

https://on.cypress.io/exec
      at https://pr8934-kq946nt66pum5lm20w5l81bcr7hecdvv.ci.cms.va.gov/__cypress/runner/cypress_runner.js:142384:26
      at tryCatcher (https://pr8934-kq946nt66pum5lm20w5l81bcr7hecdvv.ci.cms.va.gov/__cypress/runner/cypress_runner.js:13212:23)
      at Promise._settlePromiseFromHandler (https://pr8934-kq946nt66pum5lm20w5l81bcr7hecdvv.ci.cms.va.gov/__cypress/runner/cypress_runner.js:11147:31)
      at Promise._settlePromise (https://pr8934-kq946nt66pum5lm20w5l81bcr7hecdvv.ci.cms.va.gov/__cypress/runner/cypress_runner.js:11204:18)
      at Promise._settlePromise0 (https://pr8934-kq946nt66pum5lm20w5l81bcr7hecdvv.ci.cms.va.gov/__cypress/runner/cypress_runner.js:11249:10)
      at Promise._settlePromises (https://pr8934-kq946nt66pum5lm20w5l81bcr7hecdvv.ci.cms.va.gov/__cypress/runner/cypress_runner.js:11329:18)
      at _drainQueueStep (https://pr8934-kq946nt66pum5lm20w5l81bcr7hecdvv.ci.cms.va.gov/__cypress/runner/cypress_runner.js:7919:12)
      at _drainQueue (https://pr8934-kq946nt66pum5lm20w5l81bcr7hecdvv.ci.cms.va.gov/__cypress/runner/cypress_runner.js:7912:9)
      at Async.../../node_modules/bluebird/js/release/async.js.Async._drainQueues (https://pr8934-kq946nt66pum5lm20w5l81bcr7hecdvv.ci.cms.va.gov/__cypress/runner/cypress_runner.js:7928:5)
      at Async.drainQueues (https://pr8934-kq946nt66pum5lm20w5l81bcr7hecdvv.ci.cms.va.gov/__cypress/runner/cypress_runner.js:7798:14)
  From Your Spec Code:
      at Context.eval (https://pr8934-kq946nt66pum5lm20w5l81bcr7hecdvv.ci.cms.va.gov/__cypress/tests?p=tests/cypress/support/index.js:15842:13)
```

**Review by ndouglas — APPROVED** [review]

**Comment by cweagans** [comment]
Here goes nothing.

**Pull request merged**